FIX: uniformly handle string/numeric constants in multiple Python versions

klauer opened the pull request:
Referring to [greentreesnakes](https://greentreesnakes.readthedocs.io/en/latest/nodes.html#Constant), constants have changed in handling from 3.8 and on. This PR attempts to unify the two and keep compatibility - with the current feature set.

The short of it is that `ast.Constant` exists in 3.6 but is not emitted until 3.8. This means that in 3.6 and 3.7 that strings/numbers generate `Num`/`Str`, whereas in 3.8, `ast.Constant` is used for both but differentiates by way of a Python object `.value`.

For me, Python 3.7 was previously hanging on `ppci.common.CompilerError: ('Cannot do <_ast.Str object at 0x1029e8a10>', (mandelbrot.py, 31, 10))` before this PR, and works successfully after. Looks like it's not covered in the test suite currently.

klauer commented:
Did not test locally on 3.4/3.5, but now see it failing on Travis.

Rearranging the `isinstance` checks to group the constants exposed the fact that the lines including `isinstance(expr, ast.Constant)` in the current master tests were not covered - and as `ast.Constant` doesn't exist on Python 3.5 this is now failing the test. Oops.

windelbouwman commented on ppci/lang/python/python2ir.py:
```diff
@@ -526,7 +534,8 @@ def gen_call(self, expr):
     def gen_num(self, expr):
-        num = expr.n
+        # ast.Constant.value / ast.Num.n
+        num = expr.value if hasattr(expr, 'value') else expr.n
```
I'd like to modify this line, and move it to the place where the python version specifics are handled. Please modify this, such that this function does not take `expr` but gets `num` passed in.

klauer replied:
I switched it to take both - `expr` and `num`, to retain the possibility of calling `not_impl` with `expr` (0b5259a6021213489c766f240a3739e9126a0251).
If you feel strongly about it, I can switch it up exactly as you say - let me know.

windelbouwman commented on ppci/lang/python/python2ir.py:
```diff
@@ -459,12 +459,20 @@ def gen_expr(self, expr):
+                else:  # pragma: no cover
+                    self.not_impl(condition)
+            elif isinstance(expr, ast.Num):  # Python < 3.8
+                value = self.gen_num(expr)
```
Change this into:
```python
value = self.gen_num(expr.n)
```

windelbouwman commented on ppci/lang/python/python2ir.py:
```diff
@@ -459,12 +459,20 @@ def gen_expr(self, expr):
+                if isinstance(expr.value, str):
+                    value = self.gen_string_constant(expr)
+                elif isinstance(expr.value, (int, float)):
+                    value = self.gen_num(expr)
```
Change this into:
```python
value = self.gen_num(expr.value)
```

windelbouwman commented on ppci/lang/python/python2ir.py:
```diff
@@ -459,12 +459,20 @@ def gen_expr(self, expr):
+            elif hasattr(ast, 'Constant') and isinstance(expr, ast.Constant):
+                # Exists in Python 3.6+, generated in Python 3.8+
+                if isinstance(expr.value, str):
+                    value = self.gen_string_constant(expr)
```
Change this into:
```python
value = self.gen_string_constant(expr.value)
```

windelbouwman commented on ppci/lang/python/python2ir.py:
```diff
@@ -459,12 +459,20 @@ def gen_expr(self, expr):
-                value = self.gen_constant(expr)
+            elif hasattr(ast, 'Constant') and isinstance(expr, ast.Constant):
+                # Exists in Python 3.6+, generated in Python 3.8+
+                if isinstance(expr.value, str):
```
this might be handy to change into:
```python
value = expr.value
if isinstance(vale, str):
    ...
```

windelbouwman commented on ppci/lang/python/python2ir.py:
```diff
@@ -535,10 +544,10 @@ def gen_num(self, expr):
-        # TODO: are there different types of constants?
-        assert isinstance(expr.value, str)
-        data = expr.value.encode("utf8") + bytes([0])
+    def gen_string_constant(self, expr):
```
Pass in `value` directly here.

windelbouwman commented:
We check python 3.7 in travis, would it be possible to add a test case for this specific version thing? This test should pass on both 3.7 and 3.4 versions. It might even be nice to add 3.8 travis python support. The test case can be placed in `test/lang/test_python.py`

klauer commented:
Added a string constant snippet to the test suite - this should hit the changed lines. Hopefully codecov will reflect that.

Also added Python 3.8 to the test matrix (required bumping travis's distribution to bionic, as xenial only has up to 3.7).

windelbouwman commented:
The codecov is pretty broken. I guess it is fine to remove 3.4 from the build matrix, if we add 3.8 to it.

klauer commented:
Using xenial for 3.4 and bionic for the rest was the key to testing both 3.4 and 3.8 (5045301). 👍
